### DISCUSSION TITLE
OpenWRT documentation outdated?

### DISCUSSION THREAD

**Discussion by ToxicFrog** [snapcast/snapcast#1156] [Q&A]
The [installation instructions](https://github.com/badaix/snapcast/blob/develop/doc/install.md) say that "packages are available for several Linux distributions" and list OpenWRT as one of them. The [section for OpenWRT](https://github.com/badaix/snapcast/blob/develop/doc/install.md#openwrt) says to pass the ipk file itself to `opkg`, which implies that, rather than fetching it from the official repos (which would just be `opkg install snapclient`), you're meant to download the ipk manually and then install it. But the [releases](https://github.com/badaix/snapcast/releases) don't include an ipk for any architecture.

Checking `opkg` just in case, it's not in the repos for either mipsel_24kc or armv7l (which are the two openwrt devices I have). Rummaging around more, it looks like the expectation these days is to use [SnapOS](https://github.com/badaix/snapos/tree/master/openwrt) instead, which appears to be a fairly involved process.

Am I missing something, or is that now the only supported way to install snapclient on openwrt? If so, the docs should probably be updated to no longer list it as having prebuilt packages for openwrt.

**Comment by ToxicFrog** [comment]
For anyone using nixos, you can build a statically linked version that will run on openwrt arm7 thus:

```nix
let
  nixpkgs = (import <nixpkgs> {});
  pkgs = nixpkgs.pkgsCross.armv7l-hf-multiplatform.pkgsStatic;
  soxr = pkgs.soxr.overrideAttrs {
    cmakeFlags = pkgs.soxr.cmakeFlags
      ++ [ "-DWITH_OPENMP=off" ];
  };
in
  (pkgs.snapcast.override {
    pulseaudioSupport = false; libpulseaudio = null;
    avahi = null;
    inherit soxr;
  }).overrideAttrs {
    nativeBuildInputs = pkgs.snapcast.nativeBuildInputs
      ++ [ nixpkgs.pkgs.makeWrapper ];
    cmakeFlags = pkgs.snapcast.cmakeFlags
      ++ [ "-DBUILD_SERVER=off" "-DBUILD_CLIENT=on" ];
    postFixup = ''
      wrapProgram $out/bin/snapclient \
        --set ALSA_CONFIG_DIR /usr/share/alsa
      sed -Ei '
        s,/nix/store/.*/bin/bash.*,/bin/sh,
        s,/nix/store/.*snapclient-wrapped,$(dirname "$0")/.snapclient-wrapped,
      ' $out/bin/snapclient
    '';
  }
```

**Comment by kerta1n** [comment]
> Checking `opkg` just in case, it's not in the repos for either mipsel_24kc or armv7l (which are the two openwrt devices I have). Rummaging around more, it looks like the expectation these days is to use [SnapOS](https://github.com/badaix/snapos/tree/master/openwrt) instead, which appears to be a fairly involved process.

You need to build the Snapcast package for your architecture using OpenWRT's SDK (23.05.0 latest release at time of writing). Go [here](https://downloads.openwrt.org/releases/23.05.0/targets/), then go to your CPU arch, then family, and scroll all the way to the bottom. You'll find your SDK file under `Supplementary Files` (e.g. mine for the cheap supported Archer C6 was `openwrt-sdk-23.05.0-ramips-mt7621_gcc-12.3.0_musl.Linux-x86_64.tar.xz`). 

Just follow the rest of the build instructions from there and you should have the ipk file sitting in your build dir.

P.S: If someone needs a 23.05.0-supported architecture built, just respond to this as a reply
